### ISSUE TITLE
figure out list of sending addresses (aliases)

### ISSUE THREAD

**Issue by tomholub** [FlowCrypt/flowcrypt-android#80]

**Comment by tomholub** [comment]
I don't know if there is a good way to get a list of sending addresses from IMAP/SMTP. You can research. In chrome plugin, I use a recursive search function that will figure out which address an email was sent from, then search again filtering out emails sent from that address, and recursing like this until all addresses are found. It takes about 3-10 seconds.

The code:
```js
function fetch_account_aliases_from_gmail(account_email, callback, query, from_emails) {
  from_emails = from_emails || [];
  query = query || 'newer_than:1y in:sent -from:"calendar-notification@google.com" -from:"drive-shares-noreply@google.com"';
  tool.api.gmail.fetch_messages_based_on_query_and_extract_first_available_header(account_email, query, ['from'], function (headers) {
    if(headers && headers.from) {
      fetch_account_aliases_from_gmail(account_email, callback, query + ' -from:"' + tool.str.parse_email(headers.from).email + '"', from_emails.concat(tool.str.parse_email(headers.from).email));
    } else {
      callback(from_emails);
    }
  });
}

  function api_gmail_fetch_messages_based_on_query_and_extract_first_available_header(account_email, q, header_names, callback) {
    api_gmail_message_list(account_email, q, false, function (success, message_list_response) {
      if(success && typeof message_list_response.messages !== 'undefined') {
        api_gmail_fetch_messages_sequentially_from_list_and_extract_first_available_header(account_email, message_list_response.messages, header_names, callback);
      } else {
        callback(); // if the request is !success, it will just return undefined, which may not be the best
      }
    });
  }

  function api_gmail_fetch_messages_sequentially_from_list_and_extract_first_available_header(account_email, messages, header_names, callback, i) {
    // this won a prize for the most precisely named function in the hostory of javascriptkind
    i = i || 0;
    api_gmail_message_get(account_email, messages[i].id, 'metadata', function (success, message_get_response) {
      var header_values = {};
      var missing_header = false;
      if(success) { // non-mission critical - just skip failed requests
        tool.each(header_names, function (i, header_name) {
          header_values[header_name] = api_gmail_find_header(message_get_response, header_name);
          if(!header_values[header_name]) {
            missing_header = true;
          }
        });
      }
      if(!missing_header) {
        callback(header_values);
      } else if(i + 1 < messages.length) {
        api_gmail_fetch_messages_sequentially_from_list_and_extract_first_available_header(account_email, messages, header_names, callback, i + 1);
      } else {
        callback();
      }
    });
  }
```

It's not terribly complicated, but not too simple either.

**Comment by tomholub** [comment]
The simple alternative is:
 - when user opens an email in the sent folder and the email was sent from unknown email address, assume this is one of user's sending addresses.
 - later offer is as a sending address on compose screen
The downside is that the user has to manually navigate to a sent message for each sending email address they want to use. But the advantage is that it's much simpler. We could try this simpler version if you want.

**Comment by DenBond7** [comment]
I didn't quite understand this task. For now, in the app works next logic:

>When pulling sent email (sent folder): https://github.com/FlowCrypt/flowcrypt-android/issues/18

>For each sent email that you process and show in a list, collect "to", "cc" and "bcc" fields to retrieve email-name pairs. Then either insert (if no matching row) or update (if have matching row but missing name) the database records.
> 
> This is because the user may have sent these emails from before he started using FlowCrypt, or maybe sent them from other client, so they may not be in the database or the database may have missing names.
> 
> for email, name in email_name_pairs:
>     if email in db:
>         if db_row.name is null and bool(name) == true:
>             "save that person's name into the existing DB record"
>     else:
>         "save that email, name pair into DB like so: new PgpContact(email, name);" 
> All of these database records will be used to look up email addresses as the person is typing it in compose (later).

> I don't know if there is a good way to get a list of sending addresses from IMAP/SMTP. 

Maybe I can retrieve all messages from the SENT folder during setup and save the user contacts. And later will work the algorithm from issue #18

**Comment by tomholub** [comment]
This issue is about addresses I send messages from (user addresses). You can go to Gmail Settings -> Accounts and Import. Notice the "send email as" section. We need to find a way to see if user has other email addresses associated with the account, so that they can send emails from the secondary addresses as well.

**Comment by DenBond7** [comment]
For me 
https://developers.google.com/gmail/api/quickstart/android
https://developers.google.com/gmail/api/guides/alias_and_signature_settings

**Comment by tomholub** [comment]
When I was implementing it, they didn't have https://developers.google.com/gmail/api/v1/reference/users/settings/sendAs/get

But I'm not sure if we can use that in the app. One thing is oauth (not sure if this is out of scope) and another thing is that it only works for gmail, so we still need a solution for other providers.

**Comment by DenBond7** [comment]
Do other providers have aliases addresses? I don't know about that. Can you give me an example?
I thought that is only for Gmail.

**Comment by tomholub** [comment]
I'm not sure. I thought they can have more than one sending address. Eg https://www.cnet.com/how-to/how-to-send-e-mails-from-other-addresses-in-outlook-com/

But maybe we can just ignore that for non-gmail accounts, at least for now. Maybe it's not that common. If users miss it, they will let us know.

**Comment by DenBond7** [comment]
I completed the research. And I have next results:
**Terms and concepts**

- sender@gmail.com - The sender account.
- alias@outlook.com - The alias.
- receiver@example.com - The recipient account.

 Generally, we can use next options for sending email using aliases:
##############################################################################
1. The right way: need to use [Gmail API](https://developers.google.com/gmail/api/). When we will use it, we create a message with the "from" field equal one of our aliases (alias@outlook.com) and the "to" field equal "receiver@example.com".  And than call [`gmailService.users().messages().send("me", message).execute();`](https://developers.google.com/gmail/api/guides/sending). In this case, we will receive next results:
 * we have the sent message in our Gmail sent folder (sender@gmail.com);
 * we have the sent message in our outlook.com sent folder (alias@outlook.com);
 * we have the incoming message in our example.com inbox folder (receiver@example.com);

In this case, the Gmail uses SMTP settings of the Outlook account for sending an email (This is how the web version of Gmail works(in my understanding)

![image](https://user-images.githubusercontent.com/2863246/31939840-8fd9b84a-b8c4-11e7-9c2b-e145905772e4.png)

##############################################################################
2. We can only retrieve the aliases information (using [Gmail AP sendAs](https://developers.google.com/gmail/api/v1/reference/users/settings/sendAs)) and use one of the aliases to set up the "from" field. And that is all. Than we send email using standard IMAP methods. But in this case, we will receive next results:
 * we have the sent message in our Gmail sent folder (sender@gmail.com);
 * we have the incoming message in our example.com inbox folder (receiver@example.com);

But in this case, the Gmail doesn't use SMTP settings of the Outlook account for sending an email

![image](https://user-images.githubusercontent.com/2863246/31939732-38ed080c-b8c4-11e7-83da-63824f541bf0.png)

##############################################################################
3. We can retrieve the aliases information (using [Gmail API sendAs](https://developers.google.com/gmail/api/v1/reference/users/settings/sendAs)).

From the Gmail API we will receive next results:

```
{
  "sendAsEmail": string,
  "displayName": string,
  "replyToAddress": string,
  "signature": string,
  "isPrimary": boolean,
  "isDefault": boolean,
  "treatAsAlias": boolean,
  "smtpMsa": {
    "host": string,
    "port": integer,
    "username": string,
    "password": string,
    "securityMode": string
  },
  "verificationStatus": string
}
```

From this we need "smtpMsa". This structure contains information about SMTP settings of the alias. But in this case we have a problem with smtpMsa.password, because ["This is a write-only field that can be specified in requests to create or update SendAs settings; it is never populated in responses."](https://developers.google.com/gmail/api/v1/reference/users/settings/sendAs). So we must ask user about the password. When we will have the password we can send an email using SMTP settings of the alias(alias@outlook.com) and can save a copy of a sent message in the Gmail sent folder (sender@gmail.com) using IMAP.  In this case, we will receive next results:
 * we have the sent message in our Gmail sent folder (sender@gmail.com);
 * we have the sent message in our outlook.com sent folder (alias@outlook.com);
 * we have the incoming message in our example.com inbox folder (receiver@example.com);

But this option not good too.

**Comment by DenBond7** [comment]
Important information : [Gmail API quota](https://developers.google.com/gmail/api/v1/reference/quota)

**Comment by tomholub** [comment]
Looks reasonable, we can go with option #1 for Gmail accounts, and disallow this functionality for non-gmail accounts.

Please verify if our current oAuth scope is also usable for Gmail API (not just IMAP/SMTP).

**Comment by DenBond7** [comment]
> Please verify if our current oAuth scope is also usable for Gmail API (not just IMAP/SMTP).

Already did it. I can receive aliases in the app using Gmail API (because we request a [https://mail.google.com/](https://developers.google.com/gmail/api/auth/scopes) scope )

**Comment by tomholub** [comment]
That's good, I think we can proceed with that, good job researching it

--
Tom James Holub <http://holub.me/>

**Comment by DenBond7** [comment]
I completed this task. Now we can use Gmail aliases for sending messages. But I found one nuance. I do not recommend sending large files (more than 5Mb) with Gmail aliases because the [Gmail API uses standard POST request](https://developers.google.com/gmail/api/v1/reference/users/messages/send). Before sending a message, Gmail API create JSON structure with the message details (include raw data of attachment as a base64url encoded string). All of this information will be in the RAM and it can be a reason of memory leaks.
